### ISSUE TITLE
Test failure for mroi.py and Python 3.6

### ISSUE THREAD

**Issue by matthew-brett** [nipy/nipy#420]
Test failure only on Python 3.6.   Maybe a dictionary ordering error?

```
======================================================================
ERROR: Failure: IndexError (list index out of range)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nose/failure.py", line 39, in runTest
    raise self.exc_val.with_traceback(self.tb)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nose/loader.py", line 252, in generate
    for test in g():
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/tests/test_bsa.py", line 84, in test_bsa_methods
    algorithm=name)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/tests/test_bsa.py", line 37, in make_bsa_2d
    algorithm=algorithm, n_iter=100, burnin=10)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/bayesian_structural_analysis.py", line 375, in compute_landmarks
    domain, stats, threshold, smin, method)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/bayesian_structural_analysis.py", line 133, in _compute_individual_regions
    domain, stats_, threshold=threshold, smin=smin)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/hroi.py", line 113, in HROI_as_discrete_domain_blobs
    nroi = hroi_agglomeration(nroi, criterion=criterion, smin=smin)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/hroi.py", line 71, in hroi_agglomeration
    output_hroi.select_roi(output_hroi.get_id()[value > smin])
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/hroi.py", line 301, in select_roi
    SubDomains.select_roi(self, id_list)
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/mroi.py", line 705, in select_roi
    sf = [f[id] for id in id_list_pos]
  File "/home/travis/build/matthew-brett/nipy/venv/lib/python3.6/site-packages/nipy/labs/spatial_models/mroi.py", line 705, in <listcomp>
    sf = [f[id] for id in id_list_pos]
IndexError: list index out of range
```

See: https://travis-ci.org/matthew-brett/nipy/jobs/194166695#L8304

Sure enough, when running that code via the debugger:

```
nosetests nipy/labs/spatial_models/tests/test_bsa.py:test_bsa_methods --pdb
```

```
-> sf = [f[id] for id in id_list_pos]
(Pdb) print(id)
1
(Pdb) print(f)
[array([ 3.13630455,  3.00630573,  2.8102853 ,  4.03348143,  4.14206689,
        3.81243352,  2.67251382,  2.62807902,  2.38323033])]
```

@VirgileFritsch - it looks like this is your code?  Any idea about this one?

**Comment by VirgileFritsch** [comment]
According to [this page](https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-compactdict), the behavior of dictionnaries should remain the same in Python 3.6.

Unfortunately, I am not currently able to run the code above. Also, I do not remember much about the failing test.

`self` seems to be a `MROI` object with only one ROI (because popping out one feature obviously returns a list of only one element -- `f`). On the other hand, `id` represents the position of a ROI in the list of all the ROIs. It should never be equal to 1 if there is only one ROI.
Right now, I cannot tell whether the issue comes from `id` computation of whether there is a discrepancy in `self` properties (e.g. in the list of features, in the ROI identifiers, in the voxel map, ...).

**Comment by matthew-brett** [comment]
Thanks a lot for the reply.    What has changed in Python 3.6 is the ordering of the key, value pairs.   Unfortunately, it's not uncommon that code does depend on the ordering, although of course it shouldn't.  For example, lots of tests in networkx started failing for Python 3.6 : https://github.com/networkx/networkx/pull/2319.

Is there anything I can do to help you with debugging?   Sorry not to help more, but I don't know this code at all, so I'm stuck for ideas about where to start.

**Comment by VirgileFritsch** [comment]
> What has changed in Python 3.6 is the ordering of the key, value pairs.

Ok, I misinterpreted the link I sent. In that case, it is not a surprise that my code is broken.
There is at least one line where some elements are selected according to their position in a dictionnary. It should not happen often because the code heavily relies on methods to manipulate the ROI objects.

> Is there anything I can do to help you with debugging?

Nice of you, but do not worry. I have just been asked to install Python 3.6 + "lots-of-modules" on one of my lab's machine. Running nipy on it will be a great test. I will hopefully be able to fix the ROI issue afterwards.

**Comment by bthirion** [comment]
@VirgileFritsch if you cannot deal with it, let me know, I'll take a look. Thx for your time.

Everybody should however be aware that this code is dead, hence we're basically keeping a cadaver alive.

**Comment by matthew-brett** [comment]
@VirgileFritsch - any progress?  I'd love to make a release soon.

**Comment by VirgileFritsch** [comment]
Hi Matthew, I understand you need that to be done quick. I promise I will do my best to have a look at it during the weekend. Is that ok if I give you an update, say, on Monday?

@bthirion Thank a lot for your proposal. In case I do not stick to the deadline I gave above, then maybe it is worthy that you have a look.

Hope it helps!

**Comment by matthew-brett** [comment]
Virgile - that's fine, thanks, please do let me know on Monday.

**Comment by VirgileFritsch** [comment]
Sorry Matthew, I could not find the time this weekend to have a look at the issue.
This week is quite a busy week because of other projects. When do you plan to make the release?

**Comment by matthew-brett** [comment]
Ouch - can you can give me any hint to help be debug myself?   This is the feature blocking the release, I'd release now if this was fixed.

**Comment by matthew-brett** [comment]
Virgile - see https://github.com/nipy/nipy/pull/422 - is that the right fix?

**Comment by matthew-brett** [comment]
See http://bugs.python.org/issue29420.

Closed by #422 .
